src/main/scala/io/iohk/ethereum/vm/VM.scala

```diff
@@ -1,36 +1,88 @@
 package io.iohk.ethereum.vm
 
+import akka.util.ByteString
+import io.iohk.ethereum.domain.Address
 import io.iohk.ethereum.utils.Logger
 
 import scala.annotation.tailrec
 
-/**
-  * Entry point to executing a program.
-  */
-class VM extends Logger {
+class VM[W <: WorldStateProxy[W, S], S <: Storage[S]] extends Logger {
+
+  type PC = ProgramContext[W, S]
+  type PR = ProgramResult[W, S]
+  type PS = ProgramState[W, S]
 
   /**
-    * Executes a program
+    * Executes a top-level program (transaction)
     * @param context context to be executed
     * @return result of the execution
    */
-  def run[W <: WorldStateProxy[W, S], S <: Storage[S]](context: ProgramContext[W, S]): ProgramResult[W, S] = {
-    PrecompiledContracts.runOptionally(context).getOrElse {
-      val finalState = run(ProgramState(context))
-      ProgramResult[W, S](
-        finalState.returnData,
-        finalState.gas,
-        finalState.world,
-        finalState.addressesToDelete,
-        finalState.logs,
-        finalState.internalTxs,
-        finalState.gasRefund,
-        finalState.error)
+  def run(context: ProgramContext[W, S]): ProgramResult[W, S] = {
+    {
+      import context._
+      import org.spongycastle.util.encoders.Hex
+      log.trace(s"caller:  $callerAddr | recipient: $recipientAddr | gasPrice: $gasPrice | value: $value | inputData: ${Hex.toHexString(inputData.toArray)}")
     }
+
+    if (context.recipientAddr.isEmpty)
+      create(context)._1
+
+    else
+      call(context, context.recipientAddr.get)
```

Comment:
why not just match on recipient addr?

Reply:
doh 👍.. it's a leftover from something I tried earlier